docs/proposals/image-index.md

```diff
@@ -0,0 +1,153 @@
+  * All Images that the *foo* images is based on (descentants)
+* Be able to track the usage of images limited by an license by giving the image carrying the licenced
+  content a class
+* Be able to enforce policies based on classification:
+  * Allow running only *VENDOR* based images in this cluster
+* Be able to perform bulk operations based on the classification:
+  * Trigger a build for all BuildConfigs that have builder image based on *VENDOR*.
+
+### Current State
+
+The information about the Image(s) we gather is not normalized in any way. It
+does not provide any information about relationships between images, other than
+their affiliation with ImageStream. Furthermore, all of the information
+currently available is scattered throughout the entire system, and accessed
+from different parts of the system in non-unified way.
+
+This results in the tools such as pruning or `oadm top images` always build
+in-memory state of the cluster for their operation.
+
+### Goal
+
+To build efficient *Image Index API* that allows to querying the relations
+between images and also name and map these relations into human-readable form.
+
+### Known Limitations / Scope Of Work
+
+* We can only keep track of Images we know about (iow. imported into OpenShift)
+  * If somebody use an external image, we can’t map it to indexing API without importing
+    it into an Image object
+* Currently, there is no way to restrict users referencing an external images in Pods.
+  * Solving this problem goes beyond the scope of this proposal
+* This API is meant to only query the images/imagestreams not their usage in the Pods.
+* The classification is meant to be managed and used by cluster/registry administrators,
+  not by regular users
+
+## Proposed Workflow
+
+* As an admin I want to identify all *VENDOR* based images in the cluster:
+  * I create new `classification` using the *VENDOR* ImageStream(Tag)
+  * `ImageClassificationController` discovers all ancestors of this Image (or any previous
+     version of this Image) in the cluster
+  * `ImageClassificationController` applies label/annotations to the Image(s) that include
+    reference to *VENDOR*
+
+* As an admin I want to group the classifications together, so I can send "key=value" queries, for example:
+  * I want to get all images with `os=rhel7` or `vendor=redhat` or `lang=python`:
+  * For that I create a new classification group named “os” and provide an “selector” that matches the labels
+    in the target “classification” objects.
+
+
+### Proposed API Changes
+
+For the implementation, two new API types are needed:
+
+```go
+// ImageClass provides a single Image "class" a context that a
+// certain group if Images shares. For example "rhel7" or "python".
+// To create an ImageClass, you have to point it to an initial
+// or "target" image first. From that point, the ImageClassController
+// will track the history of that Image.
+type ImageClass struct{
```

Comment:
Since image will be annotated by `key=value` pair based on an instance of this class, I'd expect it to contain at least a key name (e.g. `Annotation="os"`). The value could be taken from the `Name` but in this case the value would lack a scope (the annotation) which could result in name collisions.

Or maybe both key and value could be encoded in the `Name` (e.g. `Name="os:rhel7"`).